From 9d230c174dd3202bd1d6025fc637cedcefe75db5 Mon Sep 17 00:00:00 2001
From: Sathish Ambley <sathishambley@codeaurora.org>
Date: Tue, 6 Oct 2015 10:58:12 -0700
Subject: [PATCH 3/5] msm: ADSPRPC: Add carve out for ADSP heap

Add ADSP ION heap carve out for sharing memory across
ADSP subsystem and update driver to be able to communicate
with the corresponding DSP subsystem.

Change-Id: I4dbb7529f385af03eff5cfff86a22454a3f76a79
Signed-off-by: Sathish Ambley <sathishambley@codeaurora.org>
---
 arch/arm/boot/dts/qcom/msm8916-512mb-ion.dtsi    |  7 ++++++
 arch/arm/boot/dts/qcom/msm8916-512mb-memory.dtsi |  8 ++++++-
 arch/arm/boot/dts/qcom/msm8916-ion.dtsi          |  7 ++++++
 arch/arm/boot/dts/qcom/msm8916-memory.dtsi       |  8 ++++++-
 drivers/char/adsprpc.c                           | 29 +++++++++++++++++++-----
 5 files changed, 51 insertions(+), 8 deletions(-)

diff --git a/arch/arm/boot/dts/qcom/msm8916-512mb-ion.dtsi b/arch/arm/boot/dts/qcom/msm8916-512mb-ion.dtsi
index b688a10..e1f1012 100644
--- a/arch/arm/boot/dts/qcom/msm8916-512mb-ion.dtsi
+++ b/arch/arm/boot/dts/qcom/msm8916-512mb-ion.dtsi
@@ -26,6 +26,13 @@
 			qcom,ion-heap-type = "SYSTEM_CONTIG";
 		};
 
+                qcom,ion-heap@22 { /* adsp heap */
+                        compatible = "qcom,msm-ion-reserve";
+                        reg = <22>;
+                        linux,contiguous-region = <&adsp_mem>;
+                        qcom,ion-heap-type = "DMA";
+                };
+
 		qcom,ion-heap@27 { /* QSEECOM HEAP */
 			compatible = "qcom,msm-ion-reserve";
 			reg = <27>;
diff --git a/arch/arm/boot/dts/qcom/msm8916-512mb-memory.dtsi b/arch/arm/boot/dts/qcom/msm8916-512mb-memory.dtsi
index 8103c57..fb6a74c 100644
--- a/arch/arm/boot/dts/qcom/msm8916-512mb-memory.dtsi
+++ b/arch/arm/boot/dts/qcom/msm8916-512mb-memory.dtsi
@@ -29,7 +29,7 @@
 			linux,reserve-contiguous-region;
 			linux,reserve-region;
 			linux,remove-completely;
-			reg = <0x0 0x86800000 0x0 0x04800000>;
+			reg = <0x0 0x86800000 0x0 0x02b00000>;
 			label = "modem_adsp_mem";
 		};
 
@@ -48,6 +48,12 @@
 			label = "venus_qseecom_mem";
 		};
 
+                adsp_mem: adsp_region@0 {
+                        linux,reserve-contiguous-region;
+                        reg = <0 0 0 0x1000000>;
+                        label = "adsp_mem";
+                };
+
 		audio_mem: audio_region@0 {
 			linux,reserve-contiguous-region;
 			reg = <0 0 0 0x314000>;
diff --git a/arch/arm/boot/dts/qcom/msm8916-ion.dtsi b/arch/arm/boot/dts/qcom/msm8916-ion.dtsi
index 80baf241..80f5b7c 100644
--- a/arch/arm/boot/dts/qcom/msm8916-ion.dtsi
+++ b/arch/arm/boot/dts/qcom/msm8916-ion.dtsi
@@ -34,6 +34,13 @@
 			qcom,ion-heap-type = "SECURE_DMA";
 		};
 
+                qcom,ion-heap@22 { /* adsp heap */
+                        compatible = "qcom,msm-ion-reserve";
+                        reg = <22>;
+                        linux,contiguous-region = <&adsp_mem>;
+                        qcom,ion-heap-type = "DMA";
+                };
+
 		qcom,ion-heap@27 { /* QSEECOM HEAP */
 			compatible = "qcom,msm-ion-reserve";
 			reg = <27>;
diff --git a/arch/arm/boot/dts/qcom/msm8916-memory.dtsi b/arch/arm/boot/dts/qcom/msm8916-memory.dtsi
index 3e18047..f0783fc 100644
--- a/arch/arm/boot/dts/qcom/msm8916-memory.dtsi
+++ b/arch/arm/boot/dts/qcom/msm8916-memory.dtsi
@@ -29,7 +29,7 @@
 			linux,reserve-contiguous-region;
 			linux,reserve-region;
 			linux,remove-completely;
-			reg = <0x0 0x86800000 0x0 0x05600000>;
+			reg = <0x0 0x86800000 0x0 0x02b00000>;
 			label = "modem_adsp_mem";
 		};
 
@@ -54,6 +54,12 @@
 			label = "venus_qseecom_mem";
 		};
 
+		adsp_mem: adsp_region@0 {
+			linux,reserve-contiguous-region;
+			reg = <0 0 0 0x1000000>;
+			label = "adsp_mem";
+		};
+
 		audio_mem: audio_region@0 {
 			linux,reserve-contiguous-region;
 			reg = <0 0 0 0x314000>;
diff --git a/drivers/char/adsprpc.c b/drivers/char/adsprpc.c
index 9999ff5..9f81cdd 100644
--- a/drivers/char/adsprpc.c
+++ b/drivers/char/adsprpc.c
@@ -212,6 +212,7 @@ struct fastrpc_apps {
 	struct smq_phy_page range;
 	dev_t dev_no;
 	int compat;
+	int adsp_channel;
 	spinlock_t wrlock;
 	spinlock_t hlock;
 	struct hlist_head htbl[RPC_HASH_SZ];
@@ -1692,16 +1693,19 @@ static int fastrpc_device_release(struct inode *inode, struct file *file)
 static int fastrpc_device_open(struct inode *inode, struct file *filp)
 {
 	int cid = MINOR(inode->i_rdev);
-	int err = 0, ssrcount;
+	int err = 0, ssrcount, channel;
 	struct fastrpc_apps *me = &gfa;
 
 	mutex_lock(&me->smd_mutex);
 	ssrcount = me->channel[cid].ssrcount;
 	if ((kref_get_unless_zero(&me->channel[cid].kref) == 0) ||
 		(me->channel[cid].chan == 0)) {
+		channel = gcinfo[cid].channel;
+		if (!strcmp(gcinfo[cid].subsys, "adsp") && !me->adsp_channel)
+			channel = SMD_APPS_MODEM;
 		VERIFY(err, 0 == smd_named_open_on_edge(
 					FASTRPC_SMD_GUID,
-					gcinfo[cid].channel,
+					channel,
 					&me->channel[cid].chan,
 					(void *)(uintptr_t)cid,
 					smd_event_handler));
@@ -1871,8 +1875,9 @@ static const struct file_operations fops = {
 static int __init fastrpc_device_init(void)
 {
 	struct fastrpc_apps *me = &gfa;
-	struct device_node *ion_node, *node, *pnode;
+	struct device_node *cnode, *node, *pnode;
 	struct platform_device *pdev;
+	const char *label;
 	const u32 *addr;
 	uint64_t size;
 	uint32_t val;
@@ -1897,9 +1902,9 @@ static int __init fastrpc_device_init(void)
 	if (err)
 		goto class_create_bail;
 	me->compat = (NULL == fops.compat_ioctl) ? 0 : 1;
-	ion_node = of_find_compatible_node(NULL, NULL, "qcom,msm-ion");
-	if (ion_node) {
-		for_each_available_child_of_node(ion_node, node) {
+	cnode = of_find_compatible_node(NULL, NULL, "qcom,msm-ion");
+	if (cnode) {
+		for_each_available_child_of_node(cnode, node) {
 			if (of_property_read_u32(node, "reg", &val))
 				continue;
 			if (val != ION_ADSP_HEAP_ID)
@@ -1920,6 +1925,18 @@ static int __init fastrpc_device_init(void)
 			break;
 		}
 	}
+	node = of_find_compatible_node(NULL, NULL, "qcom,smem");
+	if (cnode) {
+		for_each_available_child_of_node(cnode, node) {
+			if (of_property_match_string(node, "compatible",
+							"qcom,smd") < 0)
+				continue;
+			if (of_property_read_string(node, "label", &label))
+				continue;
+			if (!strcmp(label, "adsp"))
+				me->adsp_channel = 1;
+		}
+	}
 	for (i = 0; i < NUM_CHANNELS; i++) {
 		me->channel[i].dev = device_create(me->class, NULL,
 					MKDEV(MAJOR(me->dev_no), i),
-- 
1.8.2.1

